### PR TITLE
Expose all gRPC endpoints as JSON POST endpoints

### PR DESCRIPTION
**Pull request by swankjesse**
We use the same mechanism we were using to expose proto endpoints
as JSON endpoints.

This also changes misk-web to use httpMethod instead of
dispatchMechanism to describe server-side endpoints. These
two concepts are related but different and conflating them
was broken for gRPC.

### CODE REVIEW

File: misk/web/tabs/web-actions/src/ducks/webActions.ts [hunk 1]
```diff
@@ -79,7 +79,7 @@ export interface IWebActionAPI {
   allowedServices: string[]
   allowedCapabilities: string[]
   applicationInterceptors: string[]
-  dispatchMechanism: HTTPMethod
```
Comment (swankjesse) [inline misk/web/tabs/web-actions/src/ducks/webActions.ts:82]: this was problematic because the HTTPMethod enum didn’t really support GRPC.

**Comment by adrw** [comment]
I'm unfamiliar with gRPC, does it still have concepts like method (Get, Post, Put...) or is relying on that `HTTPMethod` enum in Web Actions tab incompatible with how we think about gRPC? We can migrate from that enum (3rd party library) and roll our own in Misk-Web that better supports how we think about gRPC.

File: misk/src/test/kotlin/misk/web/JsonForProtoEndpointsTest.kt [hunk 2]
```diff
@@ -79,14 +129,64 @@ internal class JsonForProtoEndpointsTest {
...
+      )))
+      install(WebActionModule.create<ProtoEchoShipmentToken>())
+      install(WebActionModule.create<GrpcEchoShipmentToken>())
     }
   }
 
-  class EchoShipmentToken @Inject constructor() : WebAction {
+  // NB: The server doesn't get a port until after it starts so we create the client module
+  // _after_ we start the services
+  class ClientModule(val jetty: JettyService) : KAbstractModule() {
```
Comment (acostama) [inline misk/src/test/kotlin/misk/web/JsonForProtoEndpointsTest.kt:163]: does this class still need to be here? can't see where it used..

File: misk/web/tabs/web-actions/src/ducks/webActions.ts [hunk 3]
```diff
@@ -154,7 +154,7 @@ export const WebActionInternalLabel: { [key: string]: string } = {
   "Allowed Capabilities": "allowedCapabilities",
   "Allowed Services": "allowedServices",
   "Application Interceptor": "applicationInterceptors",
-  "Dispatch Mechanism": "dispatchMechanism",
+  "Dispatch Mechanism": "httpMethod",
```
Comment (adrw) [inline misk/web/tabs/web-actions/src/ducks/webActions.ts:157]: Can we update this to "HTTP Method" as well?

**Comment by acostama** [comment]
> This also changes misk-web to use httpMethod instead of
> dispatchMechanism to describe server-side endpoints.

a little unfamiliar here but, if I have a gRPC endpoint and the same POST endpoint is there potential for ambiguity?

File: misk/src/test/kotlin/misk/web/JsonForProtoEndpointsTest.kt (hunk 2 above)
Comment (swankjesse) [reply misk/src/test/kotlin/misk/web/JsonForProtoEndpointsTest.kt:163]: ooooh good catch. Fixed!

File: misk/web/tabs/web-actions/src/ducks/webActions.ts (hunk 3 above)
Comment (swankjesse) [reply misk/web/tabs/web-actions/src/ducks/webActions.ts:157]: YES!

Comment (swankjesse) [reply misk/web/tabs/web-actions/src/ducks/webActions.ts:157]: Fixed.

**Comment by swankjesse** [comment]
@acostama there is in theory, but unlikely in practice. If you do have a collision Misk will crash. (It’s unlikely in practice because gRPC path declarations are unlikely to be what you’d pick naturally.)

**Comment by adrw** [comment]
Looks like node build failures are snapshot tests. Update them with `miskweb test-update -e` and install the CLI if you don't have it with `npm install -g @misk/cli`

File: misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt [hunk 4]
```diff
@@ -130,8 +132,20 @@ internal class WebActionFactory @Inject constructor(
    * Returns null if neither of the input action's content types were protobuf.
    */
   private fun transformActionIntoJson(action: Action): Action? {
-    var jsonVariant = action
+    // Map gRPC actions to the equivalent JSON actions. This changes the dispatch mechanism from
+    // gRPC to POST which prevents gRPC framing from being used.
+    if (action.dispatchMechanism == DispatchMechanism.GRPC && action.parameters.size == 1) {
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt:137]: what happens if it's a grpc stream?

Comment (ryanhall07) [inline misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt:137]: what does the parameter.size == 1 mean?

File: misk/src/test/kotlin/misk/web/JsonForProtoEndpointsTest.kt [hunk 5]
```diff
@@ -96,7 +159,37 @@ internal class JsonForProtoEndpointsTest {
...
+      ShippingGetDestinationWarehouseBlockingServer, WebAction {
+    @Unauthenticated
+    override fun GetDestinationWarehouse(shipment: Shipment): Warehouse {
+      return Warehouse.Builder()
+          .warehouse_token(shipment.shipment_token)
+          .build()
+    }
+  }
+
+  // TODO(jwilson): get Wire to generate this interface.
```
Comment (ryanhall07) [inline misk/src/test/kotlin/misk/web/JsonForProtoEndpointsTest.kt:172]: so at the moment you have to manually define these if you want to use http/json?

File: misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt (hunk 4 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt:137]: parameters.size won’t be 1 and it won’t get any special treatment!

Comment (swankjesse) [reply misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt:137]: gRPC streams have two parameters (inbound and outbound streams)

File: misk/src/test/kotlin/misk/web/JsonForProtoEndpointsTest.kt (hunk 5 above)
Comment (swankjesse) [reply misk/src/test/kotlin/misk/web/JsonForProtoEndpointsTest.kt:172]: no definitely not. I just need to change how Misk generates its test .proto files from that Bash script to the Wire plugin. Seemed like too much for this PR.

File: misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt (hunk 4 above)
Comment (ryanhall07) [reply misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt:137]: it seems like we could at least support server streaming with long polling

Comment (ryanhall07) [reply misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt:137]: k. a think a code comment would be helpful here

**Pull request merged**